template/registry/registry.go

```diff
@@ -0,0 +1,153 @@
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Registry []Func
+
+// Func is a function that can be used in formatting and matching contexts.
+type Func struct {
+	FormatSignature *Signature                             // Signature of the function when called in formatting context
+	MatchSignature  *Signature                             // Signature of the function when called in matching context
+	F               func(any, map[string]any) (any, error) // Function itself
+	Name            string
+	Description     string
+}
+
+// Signature is a signature of the function, i.e. what input and options are allowed.
+type Signature struct {
+	ValidateInput func(any) error // Only when input is true
+	Options       Options         // Possible options for the function
+	Input         bool            // If true, input is required
+}
+
+// Option is a possible options for the function.
+type Option struct {
+	Name        string
+	Description string
+	Default     any
+
+	ValidateValue  func(any) error // If option value is not restricted by a set of values.
+	PossibleValues []any           // If option value is restricted by a set of values.
+}
+
+type Options []Option
+
+func NewRegistry() Registry {
+	return []Func{
+		*stringRegistryF,
+		*numberRegistryF,
+	}
+}
+
+func (r Registry) Find(name string) (Func, bool) {
+	for _, fn := range r {
+		if fn.Name == name {
+			return fn, true
+		}
+	}
+
+	return Func{}, false
+}
+
+func (r *Func) Format(input any, options map[string]any) (any, error) {
+	if r.FormatSignature == nil {
+		return "", fmt.Errorf("function '%s' is not allowed to use in formatting context", r.Name)
+	}
+
+	if err := r.FormatSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return r.F(input, options)
+}
+
+func (r *Func) Match(input any, options map[string]any) (any, error) {
+	if r.MatchSignature == nil {
+		return "", fmt.Errorf("function '%s' is not allowed to use in selector context", r.Name)
+	}
+
+	if err := r.MatchSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return r.F(input, options)
+}
+
+func (s *Signature) check(input any, options map[string]any) error {
+	if s.Input && input == nil {
+		return errors.New("input is nil, but required")
+	}
+
+	if !s.Input && input != nil {
+		return errors.New("input is not allowed")
+	}
+
+	if s.Options == nil && len(options) > 0 {
```

Comment:
Is it supposed to allow an empty slice in `s.Options`?

```suggestion
	if len(s.Options) == 0 && len(options) > 0 {
```

Reply:
nope, u right